app/src/main/java/eu/kanade/tachiyomi/data/track/kavita/KavitaApi.kt

```diff
@@ -0,0 +1,191 @@
+package eu.kanade.tachiyomi.data.track.kavita
+
+import android.app.Application
+import android.content.SharedPreferences
+import eu.kanade.tachiyomi.BuildConfig
+import eu.kanade.tachiyomi.data.database.models.Track
+import eu.kanade.tachiyomi.data.track.TrackManager
+import eu.kanade.tachiyomi.data.track.model.TrackSearch
+import eu.kanade.tachiyomi.network.GET
+import eu.kanade.tachiyomi.network.POST
+import eu.kanade.tachiyomi.network.await
+import eu.kanade.tachiyomi.network.parseAs
+import eu.kanade.tachiyomi.util.lang.withIOContext
+import eu.kanade.tachiyomi.util.system.logcat
+import logcat.LogPriority
+import okhttp3.Headers
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody.Companion.toRequestBody
+import uy.kohesive.injekt.Injekt
+import uy.kohesive.injekt.api.get
+import java.security.MessageDigest
+
+class KavitaApi(private val client: OkHttpClient) {
+    var jwtToken = ""
+    private fun headersBuilder(): Headers.Builder {
+        return Headers.Builder()
+            .add("User-Agent", "Tachiyomi Kavita v${BuildConfig.VERSION_NAME}")
+            .add("Content-Type", "application/json")
+            .add("Authorization", "Bearer $jwtToken")
+    }
+
+    private fun getKavitaPreferencesApiKey(apiUrl: String): String {
+        var prefApiKey = ""
+        for (sourceId in 1..3) {
+            val sourceSuffixID by lazy {
+                val key = "${"kavita_$sourceId"}/all/1" // Hardcoded versionID to 1
+                val bytes = MessageDigest.getInstance("MD5").digest(key.toByteArray())
+                (0..7).map { bytes[it].toLong() and 0xff shl 8 * (7 - it) }
+                    .reduce(Long::or) and Long.MAX_VALUE
+            }
+            val preferences: SharedPreferences by lazy {
+                Injekt.get<Application>().getSharedPreferences("source_$sourceSuffixID", 0x0000)
+            }
+            val prefApiUrl = preferences.getString("APIURL", "")!!
+
+            if (prefApiUrl.isNotEmpty()) {
+                if (prefApiUrl == getCleanedApiUrl(apiUrl)) {
+                    prefApiKey = Injekt.get<Application>()
+                        .getSharedPreferences("source_$sourceSuffixID", 0x0000)
+                        .getString("APIKEY", "")!!
+                    break
+                }
+            }
+        }
+        return prefApiKey
+    }
+
+    var apiUrl = ""
+    private fun getCleanedApiUrl(url: String): String {
+        /*Returns "<kavita_IP-Domain>/api/" from given url,*/
+        apiUrl = "${url.split("/api/").first()}/api"
+        return apiUrl
+    }
```

Comment:
Could perhaps be simplified to this
```suggestion
    var apiUrl = ""
        set(value) {
            field = "${value.split("/api/").first()}/api"
        }
```